cmake/ranges_flags.cmake

```diff
@@ -0,0 +1,184 @@
+  endif()
+endif()
+
+# Sanitizer support: detect incompatible sanitizer combinations
+if (RANGES_ASAN AND RANGES_MSAN)
+  message(FATAL_ERROR "[range-v3 error]: AddressSanitizer and MemorySanitizer are both enabled at the same time!")
+endif()
+
+if (RANGES_MSAN AND RANGES_ENV_MACOSX)
+  message(FATAL_ERROR "[range-v3 error]: MemorySanitizer is not supported on MacOSX!")
+endif()
+
+# AddressSanitizer support
+if (RANGES_ASAN)
+  if (RANGES_ENV_MACOSX) # LeakSanitizer not supported on MacOSX
+    ranges_append_flag(RANGES_HAS_ASAN "-fsanitize=address,integer,undefined,nullability")
+  else()
+    ranges_append_flag(RANGES_HAS_ASAN "-fsanitize=address,integer,undefined,leak,nullability")
+  endif()
+  ranges_append_flag(RANGES_HAS_SANITIZE_NO_RECOVER "-fno-sanitize-recover=all")
+  ranges_append_flag(RANGES_HAS_NO_OMIT_FRAME_POINTER -fno-omit-frame-pointer)
+endif()
+
+# MemorySanitizer support
+if (RANGES_MSAN)
+  # This policy enables passing the linker flags to the linker when trying to
+  # compile the examples, which is required to successfully link MSan binaries
+  cmake_policy(SET CMP0056 NEW)
+  ranges_append_flag(RANGES_HAS_MSAN "-fsanitize=memory")
+  ranges_append_flag(RANGES_HAS_MSAN_TRACK_ORIGINS -fsanitize-memory-track-origins)
+  ranges_append_flag(RANGES_HAS_SANITIZE_RECOVER_ALL "-fno-sanitize-recover=all")
+  ranges_append_flag(RANGES_HAS_NO_OMIT_FRAME_POINTER -fno-omit-frame-pointer)
+endif()
+
+# Build types:
+if (RANGES_DEBUG_BUILD AND RANGES_RELEASE_BUILD)
+  message(FATAL_ERROR "[range-v3 error] Cannot simultaneously generate debug and release builds!")
+endif()
+
+if (RANGES_DEBUG_BUILD)
+  ranges_append_flag(RANGES_HAS_O0 -O0)
+  ranges_append_flag(RANGES_HAS_NO_INLINE -fno-inline)
+  ranges_append_flag(RANGES_HAS_STACK_PROTECTOR_ALL -fstack-protector-all)
+  ranges_append_flag(RANGES_HAS_NO_STRICT_ALIASING -fno-strict-aliasing)
+  ranges_append_flag(RANGES_HAS_G3 -g3)
+  # Clang can generate debug info tuned for LLDB or GDB
+  if (RANGES_CXX_COMPILER_CLANG)
+    if (RANGES_ENV_MACOSX)
+      ranges_append_flag(RANGES_HAS_GLLDB -glldb)
+    elseif(RANGES_ENV_LINUX)
+      ranges_append_flag(RANGES_HAS_GGDB -ggdb)
+    endif()
+  endif()
+endif()
+
+if (RANGES_RELEASE_BUILD)
+  if (NOT RANGES_ASSERTIONS)
+    ranges_append_flag(RANGES_HAS_DNDEBUG -DNDEBUG)
+  endif()
+  if (NOT RANGES_ASAN AND NOT RANGES_MSAN)
+    # The quality of ASan and MSan error messages suffers if we disable the
+    # frame pointer, so leave it enabled when compiling with either of them:
+    ranges_append_flag(RANGES_HAS_OMIT_FRAME_POINTER -fomit-frame-pointer)
+  endif()
+
+  if (RANGES_CXX_COMPILER_CLANGC2)
+    ranges_append_flag(RANGES_HAS_O2 -O2)
+  else()
+    ranges_append_flag(RANGES_HAS_OFAST -Ofast)
+  endif()
+
+  ranges_append_flag(RANGES_HAS_STRICT_ALIASING -fstrict-aliasing)
+  ranges_append_flag(RANGES_HAS_STRICT_VTABKE_POINTERS -fstrict-vtable-pointers)
+  ranges_append_flag(RANGES_HAS_DEFINE_FAST_MATH "-D__FAST_MATH__")
```

Comment:
IIRC some clang versions did not define this but libstdc++ uses it.

This was assuming that `-ffast-math` was enabled in release, but now it isn't enabled anymore for clangC2, so this is wrong.